### PR TITLE
fix(Table): preserveSelectedRowKeys should work when checkStrictly is false

### LINKED ISSUE #30482: Table 当 checkStrictly 为 false 时， preserveSelectedRowKeys 属性不生效
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/shuxingshujuzhanshi-antd4154-forked-6jc8c?file=/index.js:1669-1692)

### Steps to reproduce
选择第一页树形，然后跳到第二页选择，第一页选中的 key 没了。

### What is expected?
保留之前选中的

### What is actually happening?
之前选中的消失了

| Environment | Info |
|---|---|
| antd | 4.15.5 |
| React | lastest |
| System | mac |
| Browser | chrome |

### PR DESCRIPTION
**Pull request by linxianxi**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [x] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

fix https://github.com/ant-design/ant-design/issues/30482
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    preserveSelectedRowKeys should work when checkStrictly is false       |
| 🇨🇳 Chinese |     当  checkStrictly 为 false 时，preserveSelectedRowKeys 应该有效     |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 0cdd8da</samp>

This pull request enhances the `rowSelection` feature of the `Table` component by allowing users to preserve the selected keys across data changes. It also adds a test case for this functionality and improves the code quality of `useSelection.tsx`.

### <samp>🤖 Generated by Copilot at 0cdd8da</samp>

*  Modify the logic of generating `keyEntities` from the data to include the `preserveSelectedRowKeys` prop ([link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L146-R164))
*  Add a test case to verify the behavior of the `rowSelection` prop with `preserveSelectedRowKeys` and `checkStrictly` set to `true` and `false` respectively ([link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-c9dafe6d2a68bda33046f3f0bf89abcc077581df1c76fcb865501a7d6ec3c005R1609-R1636))
*  Reorder the import of `act`, `fireEvent`, and `render` in `Table.rowSelection.test.tsx` to follow the alphabetical order ([link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-c9dafe6d2a68bda33046f3f0bf89abcc077581df1c76fcb865501a7d6ec3c005L4-R6))
*  Move the import of `AnyObject` type in `useSelection.ts` to line 15 and remove the duplicate import on line 28 ([link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L17-R20), [link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L30))
*  Remove or add empty lines in `useSelection.ts` to improve the readability ([link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L548), [link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L576), [link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25R592), [link](https://github.com/ant-design/ant-design/pull/42784/files?diff=unified&w=0#diff-6c5f2b8a4a1061563259571e5203e5f69eac5c1a68f2f55cf100643d59738d25L596))

### CODE REVIEW

**Comment by linxianxi** [comment]
这个陈年 bug 了，这里面豆酱大佬注释太少了，有的变量都不知道啥用处 😑

**Comment by linxianxi** [comment]
@zombieJ 看下，陈年老 bug 了

**Comment by yoyo837** [comment]
> @zombieJ 看下，陈年老 bug 了

今天没上班, 改天再催催.

File: components/table/hooks/useSelection.tsx [hunk 1]
```diff
@@ -143,16 +143,25 @@ const useSelection = <RecordType extends AnyObject = any>(
...
+    }
+    const convertData = data;
+    if (preserveSelectedRowKeys) {
+      const keys = data.map((record, index) => getRowKey(record, index));
+      // remove preserveRecords that duplicate data
+      const preserveRecords = Array.from(preserveRecordsRef.current).reduce(
+        (total: RecordType[], [key, value]) => (keys.includes(key) ? total : total.concat(value)),
+        [],
+      );
+      convertData.push(...preserveRecords);
```
Comment (zombieJ) [inline components/table/hooks/useSelection.tsx:158]: 不能直接 push，这样就把 data 给改了。应该：
```tsx
let convertData = data;

if (...) {
  convertData = [...convertData, ...preserveRecords];
}
```

File: components/table/hooks/useSelection.tsx [hunk 2]
```diff
@@ -143,16 +143,25 @@ const useSelection = <RecordType extends AnyObject = any>(
...
+  const { keyEntities } = useMemo(() => {
+    if (checkStrictly) {
+      return { keyEntities: null };
+    }
+    const convertData = data;
+    if (preserveSelectedRowKeys) {
+      const keys = data.map((record, index) => getRowKey(record, index));
+      // remove preserveRecords that duplicate data
+      const preserveRecords = Array.from(preserveRecordsRef.current).reduce(
+        (total: RecordType[], [key, value]) => (keys.includes(key) ? total : total.concat(value)),
```
Comment (zombieJ) [inline components/table/hooks/useSelection.tsx:155]: 这个性能看起来有点捉急，keys 看起来可以直接搞成 map 然后 map.has 判断

**Comment by linxianxi** [comment]
@zombieJ

File: components/table/hooks/useSelection.tsx [hunk 3]
```diff
@@ -143,16 +143,29 @@ const useSelection = <RecordType extends AnyObject = any>(
...
-          }),
-    [data, getRowKey, checkStrictly, childrenColumnName],
-  );
+  const { keyEntities } = useMemo(() => {
+    if (checkStrictly) {
+      return { keyEntities: null };
+    }
+    let convertData = data;
+    if (preserveSelectedRowKeys) {
+      const keysMap = new Map();
```
Comment (yoyo837) [inline components/table/hooks/useSelection.tsx:152]: keysMap.set(key, true); 是不是直接用Set好一点?

**Comment by zombieJ** [comment]
sync with master~

**Comment by linxianxi** [comment]
> sync with master~

done

File: components/table/__tests__/Table.rowSelection.test.tsx [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import React from 'react';
 import type { TableProps } from '..';
 import Table from '..';
-import { fireEvent, render, act } from '../../../tests/utils';
-import ConfigProvider from '../../config-provider';
+import { act, fireEvent, render } from '../../../tests/utils';
```
Comment (zombieJ) [inline components/table/__tests__/Table.rowSelection.test.tsx:4]: 这居然都会排序...

Comment (yoyo837) [reply components/table/__tests__/Table.rowSelection.test.tsx:4]: 要不要整体prettier一遍

File: components/table/hooks/useSelection.tsx [hunk 5]
```diff
@@ -143,16 +143,25 @@ const useSelection = <RecordType extends AnyObject = any>(
...
+    }
+    let convertData = data;
+    if (preserveSelectedRowKeys) {
+      const keysSet = new Set(data.map((record, index) => getRowKey(record, index)));
+      // remove preserveRecords that duplicate data
+      const preserveRecords = Array.from(preserveRecordsRef.current).reduce(
+        (total: RecordType[], [key, value]) => (keysSet.has(key) ? total : total.concat(value)),
+        [],
+      );
+      convertData = [...convertData, ...preserveRecords];
```
Comment (zombieJ) [inline components/table/hooks/useSelection.tsx:158]: 这里有点问题，`convertData` 和 `preserveRecords` 会同时存在相同的 record。如果动态删掉了 children，这个合并后的 `data` 还会是老的

**Comment by yoyo837** [comment]
要发版了, 快改改

**Comment by linxianxi** [comment]
@zombieJ 看一下，应该是没问题的

**Comment by yoyo837** [comment]
> @zombieJ 看一下，应该是没问题的

看看这个 review https://github.com/ant-design/ant-design/pull/42784#discussion_r1220996943

**Comment by zombieJ** [comment]
> @zombieJ 看一下，应该是没问题的

dataSource: [{ key: 1, children: [{ key: 2 }] }]

变为：

dataSource: [{ key: 1 }]

这个时候 cache 放后面会替换掉前面那个，变成 `[{ key: 1, children: [{ key: 2 }] }]`

**Comment by linxianxi** [comment]
> > @zombieJ 看一下，应该是没问题的
> 
> dataSource: [{ key: 1, children: [{ key: 2 }] }]
> 
> 变为：
> 
> dataSource: [{ key: 1 }]
> 
> 这个时候 cache 放后面会替换掉前面那个，变成 `[{ key: 1, children: [{ key: 2 }] }]`

reduce 里面已经把 cache 里跟 data 相同的 key去掉了，所以不会有这个问题，你再好好看下对不对

**Review by zombieJ — APPROVED** [review]

**Pull request merged**